frontend/src/react-auth0-spa.tsx

```diff
@@ -0,0 +1,124 @@
+// This file creates the Auth0Client and gives access to various hooks which can
+// be used in other components.
+// This file is provided by Auth0:
+// https://github.com/auth0/auth0-spa-js/issues/39#issuecomment-505901626
+
+import React, { useState, useEffect, useContext } from 'react';
+import createAuth0Client, {
+  PopupLoginOptions,
+  RedirectLoginResult,
+  getIdTokenClaimsOptions,
+  IdToken,
+  RedirectLoginOptions,
+  GetTokenSilentlyOptions,
+  GetTokenWithPopupOptions,
+  LogoutOptions,
+  Auth0ClientOptions,
+  Auth0Client,
+} from '@auth0/auth0-spa-js';
+
+interface Auth0Context {
+  isAuthenticated: boolean;
+  user: any;
+  loading: boolean;
+  popupOpen: boolean;
+  loginWithPopup(options: PopupLoginOptions): Promise<void>;
+  handleRedirectCallback(): Promise<RedirectLoginResult>;
+  getIdTokenClaims(o?: getIdTokenClaimsOptions): Promise<IdToken>;
+  loginWithRedirect(o: RedirectLoginOptions): Promise<void>;
+  getTokenSilently(o?: GetTokenSilentlyOptions): Promise<string | undefined>;
+  getTokenWithPopup(o?: GetTokenWithPopupOptions): Promise<string | undefined>;
+  logout(o?: LogoutOptions): void;
+}
+interface Auth0ProviderOptions {
+  children: React.ReactElement;
+  onRedirectCallback?(result: RedirectLoginResult): void;
+}
+
+const DEFAULT_REDIRECT_CALLBACK = () =>
+  window.history.replaceState({}, document.title, window.location.pathname);
+
+export const Auth0Context = React.createContext<Auth0Context | null>(null);
+export const useAuth0 = () => useContext(Auth0Context)!;
+export const Auth0Provider = ({
+  children,
+  onRedirectCallback = DEFAULT_REDIRECT_CALLBACK,
+  ...initOptions
+}: Auth0ProviderOptions & Auth0ClientOptions) => {
+  const [isAuthenticated, setIsAuthenticated] = useState(false);
+  const [user, setUser] = useState();
+  const [auth0Client, setAuth0] = useState<Auth0Client>();
+  const [loading, setLoading] = useState(true);
+  const [popupOpen, setPopupOpen] = useState(false);
+
+  useEffect(() => {
+    const initAuth0 = async () => {
+      const auth0FromHook = await createAuth0Client(initOptions);
+      setAuth0(auth0FromHook);
+
+      if (window.location.search.includes('code=')) {
+        const { appState } = await auth0FromHook.handleRedirectCallback();
+        onRedirectCallback(appState);
+      }
+
+      const checkIsAuthenticated = await auth0FromHook.isAuthenticated();
+
+      setIsAuthenticated(checkIsAuthenticated);
+
+      if (checkIsAuthenticated) {
+        const checkUser = await auth0FromHook.getUser();
+        setUser(checkUser);
+      }
+
+      setLoading(false);
+    };
+    initAuth0();
+    // eslint-disable-next-line
+  }, []);
+
+  const loginWithPopup = async (o: PopupLoginOptions) => {
+    setPopupOpen(true);
+    try {
+      await auth0Client!.loginWithPopup(o);
+    } catch (error) {
+      console.error(error);
```

Comment:
Not sure what people's opinions are on the Github Actions warning being thrown here @AdrianBorg @tjanowsk ?

Reply:
Yeah I was going to ask about this. The only reason I left it here is because it was how it was shipped with Auth0.

Reply:
According to Laurie leaving console statements is a no-no.
We probably want to give meaningful feedback to the user anyway and most of them don't look at the console.
One possibility is to use `toastr` to display the error message, I used it for the grad training project and it's quite nice.
It's not necessarily part of this PR though.

Reply:
I reckon we can leave this in for now. A console error is better than a console log, which implies the code hasn't been tidied up